### PR TITLE
[clang][rtsan] Introduce realtime sanitizer codegen and driver

### PR DESCRIPTION
**Pull request by cjappl**
Introduce the `-fsanitize=realtime` flag in clang driver

Plug in the RealtimeSanitizer PassManager pass in Codegen, and attribute a function based on if it has the `[[clang::nonblocking]]` function effect.

### CODE REVIEW

**Comment by cjappl** [comment]
I was not able to figure out a way to separate the CodeGen from the driver, seeing as the codegen only happens when the sanitizer is enabled. This means to write unit tests for the CodeGen, I needed the -fsanitize=realtime flag. If someone has advice on how to get around this problem I'm more than happy to break this up if it is too big!

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (cjappl) [inline clang/lib/CodeGen/BackendUtil.cpp:1000]: I am completely not sure where in the pipeline this pass should live. Any advice from someone with more experience would be great. This was chosen to match other passes of sanitizers around it.

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (cjappl) [inline clang/lib/CodeGen/CodeGenFunction.cpp:852]: @dougsonos , would love some eyes on this to make sure this makes sense to you!

File: clang/lib/Driver/SanitizerArgs.cpp
Comment (cjappl) [inline clang/lib/Driver/SanitizerArgs.cpp:555]: Sorry, a little clang-tidy cruft here, the only thing changed functionally was the addition of the realtime sanitizer pairs on line 560

**Comment by cjappl** [comment]
Pinging for a possible review:

@Sirraide @AaronBallman @MaskRay @vitalybuka @dougsonos 

Co-Author @davidtrevelyan

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (dougsonos) [reply clang/lib/CodeGen/CodeGenFunction.cpp:852]: `getFunctionEffects()` is cheap, but when investigating the performance regression introduced by effects, it turned out to be worthwhile to test `Context.hasAnyFunctionEffects()` first, to optimize for the case of no effects ever having been seen.

Comment (cjappl) [reply clang/lib/CodeGen/CodeGenFunction.cpp:855]: Excellent, added that check in there.

**Comment by cjappl** [comment]
(Sorry for the "double ping", I just got commit powers, so adding people to the review explicitly)

File: clang/docs/RealtimeSanitizer.rst
Comment (MaskRay) [inline clang/docs/RealtimeSanitizer.rst:76]: While `clang` can be used for a compile action, when performing a link action, `clang++` should be used to ensure libc++/libstdc++ is linked.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (vitalybuka) [reply clang/lib/CodeGen/BackendUtil.cpp:994]: This location looks OK to me.
For other sanitizers, the earlier pass is inserted, the more precise is bug detection, but binary will be slower.

If we insert late, like this one, optimization passes may optimize out a bug.

It would be nice to detect those bugs present in source code, but the cost is too high.

For RTSan, I think, users even will not even care about non-rt-friendly source code, if compiler is able to optimize it out.

So late, like here, is appropriate.

**Review by vitalybuka — APPROVED** [review]
LGTM, but please get approvals from other responded reviewers

File: clang/test/Driver/rtsan.c
Comment (MaskRay) [inline clang/test/Driver/rtsan.c:1]: Driver tests should only use `-###` to test CC1 options. We don't test different `-O` combinations. It's not useful to test `-flto`

test/CodeGen tests `%clang_cc1`

We generally don't add end-to-end tests.

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (MaskRay) [inline clang/lib/CodeGen/CodeGenFunction.cpp:848]: remove braces per https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

**Review by MaskRay — APPROVED** [review]

File: clang/test/Driver/rtsan.c
Comment (cjappl) [reply clang/test/Driver/rtsan.c:1]: The basis of this test file were very similar ones for asan and tsan:
https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/tsan.c
https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/asan.c

So I thought it was required for rtsan as well! Do you recommend I delete this test, or move it elsewhere? I defer to your expertise.

 I will fix the tests in Codegen to use clang_cc1

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (cjappl) [reply clang/lib/CodeGen/BackendUtil.cpp:994]: Thanks Vitaly! Helpful explanation

**Comment by cjappl** [comment]
Weekly ping of outstanding reviewers @Sirraide @cor3ntin @AaronBallman and @dougsonos.

Thanks in advance for any feedback/approvals before this goes in!

File: clang/test/CodeGen/rtsan_insert_at_entry.c
Comment (MaskRay) [inline clang/test/CodeGen/rtsan_insert_at_entry.c:1]: `*-unknown-linux-gnu` is more popular.

`none` could also be used for the OS component and is slightly misleading.

File: clang/test/CodeGen/rtsan_insert_at_exit.c
Comment (MaskRay) [inline clang/test/CodeGen/rtsan_insert_at_exit.c:3]: End all full sentence comments with a `.`

File: clang/test/Driver/rtsan.c
Comment (MaskRay) [reply clang/test/Driver/rtsan.c:1]: Thx for letting me know the existing asan.c/tsan. I'll try to clean up the two tests.

For this driver test which will use `-###`, `-O` is entirely uninteresting.

Sometimes `-O1` or above may be interesting for CodeGen tests but probably not for this.

File: clang/test/CodeGen/rtsan_insert_at_exit.c
Comment (MaskRay) [inline clang/test/CodeGen/rtsan_insert_at_exit.c:1]: this should be merged with `at_entry.c`.

Comment (cjappl) [reply clang/test/CodeGen/rtsan_insert_at_exit.c]: Fixed all of your latest round in the latest commit

File: clang/test/Driver/rtsan.c
Comment (cjappl) [reply clang/test/Driver/rtsan.c]: Got rid of this test completely, thanks for the help understanding all this. I'll mark this as resolved

**Comment by cjappl** [comment]
(investigating this issue right now, I'm unfamiliar with the fuzzer, so I'm trying to figure out if this is my failure or not. Will keep people posted)

**Comment by cjappl** [comment]
So it's happening in a pass away from the Realtime sanitizer pass, the top of the stack says:

```
3.	Running pass 'Function Pass Manager' on module '/home/b/sanitizer-aarch64-linux-fuzzer/build/llvm-project/compiler-rt/lib/ubsan_minimal/ubsan_minimal_handlers.cpp'.
4.	Running pass 'AArch64 Instruction Selection' on function '@_ZL12decorate_msgPcm'
 #0 0x0000b8be034b8d04 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/b/sanitizer-aarch64-linux-fuzzer/build/llvm_build0/./bin/clang+++0x6118d04)
```

I'm not sure how this change would have affected the AArch64 Instruction Selection pass, still investigating.

**Comment by cjappl** [comment]
Out of an abundance of caution I have reverted this. I am still unsure of what in my code could have caused this failure.

 If anyone out there has a bright idea please let me know. I will monitor the builds for a bit and ensure they are back to green

**Comment by cjappl** [comment]
Tests continue to fail after reverting this: https://lab.llvm.org/buildbot/#/builders/159/builds/4474

I am trying to bisect locally, I will likely reland this with no modifications tomorrow if it continues to look like this did not contribute to the build failures.

**Comment by Sirraide** [comment]
I haven’t looked too much into this, but yeah, this sounds like it might have just been an unrelated CI failure

**Comment by cjappl** [comment]
Thanks for taking a look @Sirraide , greatly appreciated.

**Comment by cjappl** [comment]
At least 3 more failures in this way since this reversion:
https://lab.llvm.org/buildbot/#/builders/159/builds/4477
https://lab.llvm.org/buildbot/#/builders/159/builds/4478
https://lab.llvm.org/buildbot/#/builders/159/builds/4479


I am going to re-land this tomorrow when I wake up

**Comment by cjappl** [comment]
Relanded as f77e8f765e42

**Pull request merged**
